src/main/java/com/redhat/rhjmc/containerjfr/core/templates/LocalStorageTemplateService.java

```diff
@@ -0,0 +1,258 @@
+
+import org.openjdk.jmc.common.unit.IConstrainedMap;
+import org.openjdk.jmc.common.unit.SimpleConstrainedMap;
+import org.openjdk.jmc.common.unit.UnitLookup;
+import org.openjdk.jmc.flightrecorder.configuration.events.EventOptionID;
+import org.openjdk.jmc.flightrecorder.controlpanel.ui.configuration.model.xml.XMLModel;
+import org.openjdk.jmc.flightrecorder.controlpanel.ui.model.EventConfiguration;
+
+import com.redhat.rhjmc.containerjfr.core.FlightRecorderException;
+import com.redhat.rhjmc.containerjfr.core.sys.Environment;
+import com.redhat.rhjmc.containerjfr.core.sys.FileSystem;
+
+public class LocalStorageTemplateService extends AbstractTemplateService
+        implements MutableTemplateService {
+
+    public static final String TEMPLATE_PATH = "CONTAINER_JFR_TEMPLATE_PATH";
+
+    private final FileSystem fs;
+    private final Environment env;
+
+    public LocalStorageTemplateService(FileSystem fs, Environment env) {
+        this.fs = fs;
+        this.env = env;
+    }
+
+    @Override
+    protected TemplateType providedTemplateType() {
+        return TemplateType.CUSTOM;
+    }
+
+    @Override
+    public void addTemplate(InputStream templateStream)
+            throws InvalidXmlException, InvalidEventTemplateException, IOException {
+        if (!env.hasEnv(TEMPLATE_PATH)) {
+            throw new IOException(
+                    String.format(
+                            "Template directory does not exist, must be set using environment variable %s",
+                            TEMPLATE_PATH));
+        }
+        Path dir = fs.pathOf(env.getEnv(TEMPLATE_PATH));
+        if (!fs.exists(dir) || !fs.isDirectory(dir) || !fs.isReadable(dir) || !fs.isWritable(dir)) {
+            throw new IOException(
+                    String.format(
+                            "Template directory %s does not exist, is not a directory, or does not have appropriate permissions",
+                            dir.toString()));
+        }
+        try (templateStream) {
+            Document doc =
+                    Jsoup.parse(
+                            templateStream, StandardCharsets.UTF_8.name(), "", Parser.xmlParser());
+            Elements els = doc.getElementsByTag("configuration");
+            if (els.isEmpty()) {
+                throw new InvalidXmlException("Document did not contain \"configuration\" element");
+            }
+            if (els.size() > 1) {
+                throw new InvalidXmlException(
+                        "Document contains multiple \"configuration\" elements");
+            }
+            Element configuration = els.first();
+            if (!configuration.hasAttr("label")) {
+                throw new InvalidXmlException(
+                        "Configuration element did not have \"label\" attribute");
+            }
+            // side effect of validation, throws ParseException or IllegalArgumentException if
+            // invalid
+            EventConfiguration.createModel(doc.toString());
```

Comment:
The setup parses through the xml twice, once with jsoup and once with the Java API, right? Could the XMLModel generated here be reused to accomplish what's needed for addTemplate? Are the checks made before this really needed? EventConfiguration is able to take an InputStream as well.